### PR TITLE
feat: auto-split an element by silence detection

### PR DESCRIPTION
**Pull request by yuto-trd**
## Summary

Adds silence-based auto-split for timeline elements:

- **`SilenceDetector`** (new, `Beutl.Engine.Audio`) — pure chunk-based silence detection over `WaveformChunk` min/max pairs. Tuning goes through a `SilenceDetectionOptions` readonly record struct (`ThresholdDb` / `MinSilenceDuration` / `Padding`) so future knobs are non-breaking.
- **`ISilenceSplitService` / `ElementSilenceSplitService`** (new, `Beutl.Editor`) — splits one or more elements at every silence-region boundary in a single history transaction, optionally deleting the silent pieces. It reuses `ElementStructureService`'s now-`internal` `SplitCore`/`DeleteCore` so the whole batch folds into one undo entry. `IElementStructureService` is untouched — this is **additive, not a breaking change** (same dedicated-service pattern as `IElementGapService` #2066 and `IElementSlipService` #2067, per design review).
- **UI**: Scene → Element menu gains "Auto Split by Silence" with a parameter prompt (threshold dB, min silence, padding, mode), wired through `TimelineTabViewModel.AutoSplitBySilenceAsync`.

Mode 3 (split + delete + ripple-shift of survivors) is deferred pending the Ripple feature (#2061); modes 1 (split-only) and 2 (split + delete silence) ship here.

## Tests

- `SilenceDetectorTests` — 18+ cases: threshold / min-duration / padding / empty / all-silent / missing-chunk-as-silent / padding-collapse / argument validation.
- `ElementSilenceSplitServiceTests` — split-only, split+delete, region-overlapping-start, whole-element-silent, no-op-no-commit, constructor null guard (via `SceneHistoryHarness`).
- Existing `ElementStructureServiceTests` stay green after the `SplitCore`/`DeleteCore` extraction.
- Local gates: `dotnet build Beutl.slnx` 0 errors; 36 relevant tests green; `dotnet format --verify-no-changes` clean.

## Review

- Pre-PR design review (2-pass): pass 1 flagged the original breaking `IElementStructureService.SplitBySilence` addition; reworked into the dedicated `ISilenceSplitService` + `SilenceDetectionOptions`; pass 2 **approved**.

## Manual verification

- Linked-video sync-split and the real-time waveform preview highlight are UI surfaces verified manually (documented in the commit body).

Refs: Project #9 board item「無音検出による自動カット」

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added **Auto Split by Silence** to the scene menu for the selected audio element, with configurable threshold (dB), minimum silence duration, padding, and mode (split only vs split and delete silence).
  * Performs automatic waveform-based silence detection and splits/deletes accordingly, with localized prompts and completion messaging.
* **Bug Fixes**
  * Improved resilience for edge cases (missing/invalid selections, no silence detected, and cancellation), ensuring safe, consistent outcomes.
* **Tests**
  * Added unit tests covering silence detection, waveform aggregation, split/delete behavior, boundary conditions, and single-commit history handling.

### CODE REVIEW

File: src/Beutl.Editor.Components/TimelineTab/ViewModels/TimelineTabViewModel.cs [hunk 1]
```diff
@@ -1119,6 +1121,44 @@ public void RazorSplitAt(TimeSpan time, bool acrossAllLayers)
...
+        {
+            chunks.Add(chunk);
+        }
+
+        IReadOnlyList<SilenceRegion> regions = SilenceDetector.Detect(chunks, duration, chunkCount, options);
+        if (regions.Count == 0) return;
+
+        EditorContext.GetRequiredService<ISilenceSplitService>()
+            .SplitBySilence(Scene, [element], regions, mode);
+    }
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl.Editor.Components/TimelineTab/ViewModels/TimelineTabViewModel.cs]: Confirmed — this was a real coordinate-space bug, fixed in 8045b2db8. SplitBySilence compares regions against Element.Start/Length (scene-timeline space), while the detector returns 0-based element-local times, so any element with a non-zero Start split at the wrong absolute positions (or no-oped). The view model now offsets detected regions by element.Start before calling the service; the detector stays pure and the service contract is scene-timeline. Two regression tests added: a [10,30s] element split correctly by scene-timeline regions, and raw element-local regions producing the old no-op.

File: src/Beutl.Engine/Audio/SilenceDetector.cs [hunk 2]
```diff
@@ -0,0 +1,104 @@
+﻿using Beutl.Engine;
+
+namespace Beutl.Audio;
+
+// Start/End are times relative to the source element's TimeRange.Start and already had
+// Padding subtracted from both ends, so splitting or deleting [Start, End] does not clip
+// the tail of the content on either side.
+public readonly record struct SilenceRegion(TimeSpan Start, TimeSpan End)
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl.Engine/Audio/SilenceDetector.cs]: Fixed in 8045b2db8 — the SilenceRegion doc comment now states times are 0-based within the analyzed waveform (not scene-timeline / not TimeRange.Start-relative).

File: src/Beutl.Editor/Services/ISilenceSplitService.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace Beutl.Editor.Services;
+
+/// <summary>
+/// Splits one or more <see cref="Element"/>s at silence-region boundaries, owning the
+/// history commit. <see cref="SplitBySilence"/> runs the whole batch (split of every
+/// target at every boundary, plus the optional delete of the silence pieces) and commits
+/// one undo entry only when the scene actually changes. Ripple-shift of the surviving
+/// pieces is not performed here — it is deferred pending the dedicated Ripple feature.
+/// </summary>
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl.Editor/Services/ISilenceSplitService.cs:12]: Fixed in 8045b2db8 — SplitBySilence's XML doc now specifies the contract: silenceRegions are scene-timeline coordinates (the same space as Element.Start), and callers must offset detector-local regions accordingly.

File: src/Beutl/Views/MainView.axaml.InitializeMenuBar.cs [hunk 4]
```diff
@@ -138,6 +141,84 @@ private async void OnDeleteElement()
...
+                Value = 100m,
+                Increment = 50m,
+                FormatString = "F0",
+                MinWidth = 220,
+            };
+            var modeCombo = new ComboBox
+            {
+                ItemsSource = new[] { "Split only", "Split + delete silence" },
+                SelectedIndex = 0,
+                MinWidth = 220,
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl/Views/MainView.axaml.InitializeMenuBar.cs:186]: Fixed in 8045b2db8 — dialog labels and mode names moved to Strings.resx / Strings.ja.resx (6 new EN/JA keys), consistent with the rest of MainView.

File: src/Beutl/Views/MainView.axaml.InitializeMenuBar.cs [hunk 5]
```diff
@@ -138,6 +141,84 @@ private async void OnDeleteElement()
...
+                    new TextBlock { Text = "Threshold (dB)" },
+                    thresholdBox,
+                    new TextBlock { Text = "Min silence duration (ms)" },
+                    minSilenceBox,
+                    new TextBlock { Text = "Padding (ms)" },
+                    paddingBox,
+                    new TextBlock { Text = "Mode" },
+                    modeCombo,
+                },
+            };
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl/Views/MainView.axaml.InitializeMenuBar.cs:204]: Fixed in 8045b2db8 — dialog labels and mode names moved to Strings.resx / Strings.ja.resx (6 new EN/JA keys), consistent with the rest of MainView.

File: src/Beutl.Editor.Components/TimelineTab/ViewModels/TimelineTabViewModel.cs [hunk 6]
```diff
@@ -1119,6 +1121,44 @@ public void RazorSplitAt(TimeSpan time, bool acrossAllLayers)
...
+
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl.Editor.Components/TimelineTab/ViewModels/TimelineTabViewModel.cs]: Fixed in 8045b2db8 — the waveform analysis (including GetWaveformChunksAsync) is now wrapped in try/catch: OperationCanceledException is ignored, other failures log via _logger.LogError and surface a NotificationService.ShowError message, following the existing PasteCore/HandleDuplicateException pattern.

File: src/Beutl.Editor.Components/TimelineTab/ViewModels/TimelineTabViewModel.cs [hunk 7]
```diff
@@ -1119,6 +1121,44 @@ public void RazorSplitAt(TimeSpan time, bool acrossAllLayers)
...
+        int chunkCount = Math.Clamp((int)(duration.TotalSeconds * 20), 200, 8000);
+
+        var chunks = new List<WaveformChunk>();
+        await foreach (WaveformChunk chunk in provider.GetWaveformChunksAsync(chunkCount, SamplesPerChunk, ThumbnailCacheService.Instance, cancellationToken))
+        {
+            chunks.Add(chunk);
+        }
+
+        IReadOnlyList<SilenceRegion> regions = SilenceDetector.Detect(chunks, duration, chunkCount, options);
+        if (regions.Count == 0) return;
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl.Editor.Components/TimelineTab/ViewModels/TimelineTabViewModel.cs]: Fixed in 8045b2db8 — silent no-ops removed: no audio element selected shows a warning notification, zero detected regions / zero splits shows an info notification, and success reports split/deleted counts, matching the existing DuplicateSelectedElements notification pattern.

File: src/Beutl/Views/MainView.axaml.InitializeMenuBar.cs [hunk 8]
```diff
@@ -138,6 +141,84 @@ private async void OnDeleteElement()
...
+
+            if (await dialog.ShowAsync() == ContentDialogResult.Primary)
+            {
+                var options = new SilenceDetectionOptions(
+                    (double)(thresholdBox.Value ?? -40m),
+                    TimeSpan.FromMilliseconds((double)(minSilenceBox.Value ?? 300m)),
+                    TimeSpan.FromMilliseconds((double)(paddingBox.Value ?? 100m)));
+                var mode = (SilenceSplitMode)Math.Max(0, modeCombo.SelectedIndex);
+                await timeline.AutoSplitBySilenceAsync(options, mode);
+            }
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl/Views/MainView.axaml.InitializeMenuBar.cs:233]: Fixed in 8045b2db8 — OnAutoSplitBySilence now guards the await with try/catch + ex.Handle() + NotificationService.ShowError, same as OnExportProject.

File: src/Beutl.Editor.Components/TimelineTab/ViewModels/TimelineTabViewModel.cs [hunk 9]
```diff
@@ -1453,6 +1455,90 @@ public void RazorSplitAt(TimeSpan time, bool acrossAllLayers)
...
+        {
+            NotificationService.ShowWarning(Strings.AutoSplitBySilence, MessageStrings.AutoSplitBySilence_NoSilenceDetected);
+            return;
+        }
+
+        SilenceSplitOutcome outcome;
+        try
+        {
+            const int SamplesPerChunk = 4096;
+            int chunkCount = Math.Clamp((int)(duration.TotalSeconds * 20), 200, 8000);
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl.Editor.Components/TimelineTab/ViewModels/TimelineTabViewModel.cs]: Fixed in 66e611882. `src/Beutl.Editor.Components/TimelineTab/Services/SilenceWaveformAnalysis.cs:43` now requests complete chunk spans with `samplesPerChunk: int.MaxValue` and bypasses preview waveform caches, so audible samples after the first 4,096 are included in the deletion decision. `tests/Beutl.UnitTests/Editor/Services/SilenceWaveformAnalysisTests.cs:33` covers the complete, uncached request.

File: src/Beutl.Editor.Components/TimelineTab/ViewModels/TimelineTabViewModel.cs [hunk 10]
```diff
@@ -1453,6 +1455,90 @@ public void RazorSplitAt(TimeSpan time, bool acrossAllLayers)
...
+            string.Format(MessageStrings.AutoSplitBySilence_Completed, outcome.SplitCount, outcome.DeletedCount));
+    }
+
+    private static IThumbnailsProvider? FindAudioThumbnailsProvider(Element element)
+    {
+        foreach (EngineObject obj in element.Objects)
+        {
+            if (obj is IThumbnailsProvider { ThumbnailsKind: ThumbnailsKind.Audio } provider)
+                return provider;
+        }
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl.Editor.Components/TimelineTab/ViewModels/TimelineTabViewModel.cs]: Fixed in 66e611882. `src/Beutl.Editor.Components/TimelineTab/ViewModels/TimelineTabViewModel.cs:1469` now collects every enabled audio provider, and `src/Beutl.Editor.Components/TimelineTab/Services/SilenceWaveformAnalysis.cs:38` analyzes all of them and conservatively sums their peak bounds before silence detection. `tests/Beutl.UnitTests/Editor/Services/SilenceWaveformAnalysisTests.cs:14` and `:49` cover provider selection and multi-source classification.
